Infinite Scroll: Load registered inline scripts correctly.

pento opened the pull request:
Infinite scroll currently doesn't execute any registered inline scripts. This causes issues for anything that loads `@wordpress/*` scripts: these scripts invariably depend on `lodash`, and `lodash` has an inline script to ensure it's registered in noConflict mode. Without this inline script, `@wordpress/*` scripts will be unable to use it.

This PR fixes it by executing inline scripts at the appropriate time.

Additionally, this PR fixes an issue with the map block, which causes it to only run its init functions on initial page load, which is incompatible with infinite scroll.

Fixes #15024.
Fixes #15025.

#### Changes proposed in this Pull Request:
* Add inline script support to infinite scroll.
* Fix map block infinite scroll compatibility.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Adds a new feature to infinite scroll.
* Fixes a bug in the map block.

#### Testing instructions:

* Follow the testing instructions in #15024.
* Follow the testing instructions in #15025.

#### Proposed changelog entry for your changes:
* Infinite Scroll: Added support for running inline scripts.
* Map Block: Fixed infinite scroll compatibility.

pento committed:
Set lodash to noconflict mode after loading it.

pento committed:
If the document is already loaded (eg, in infinite scroll), the window load even will never fire.

pento committed:
Add inline script support to infinite scroll.

pento committed:
Switch to using domReady for loading the map block view script.

scruffian commented on modules/infinite-scroll/infinity.js:
```diff
@@ -411,6 +423,20 @@
+
+							document.getElementsByTagName( elementToAppendTo )[ 0 ].appendChild( after );
+						};
+					}
```
This code block appears 3 times in this file so it might be worth refactoring to a new function

pento replied:
Good point, done in 9193a5dd6.

scruffian commented:
I tested this and it doesn't seem to change anything. I'll dig in later if I get time...

scruffian commented:
I had to add `SCRIPT_DEBUG` to get this working. LGTM

pento committed:
Merge the all the inline script tag creation into a helper function.

jeherve reviewed (commented):
It works well for the Map block, but does not seem to solve the issue for the Slideshow block on my end. 

Should we update our use of `@wordpress/dom-ready` for all the block that were touched in d0453da32b31f2657020892fe7464d9f8f481087?

jeherve reviewed (dismissed):
As an alternative, we could merge this PR for now and address the other blocks later. I'll approve this PR so you can merge it if you'd prefer that approach. 

Just a quick note that before you do merge that PR, you'd have to update the PR description so it does not auto-close #15024.

aaronrobertshaw commented:
This tested well for me. After scrolling, the posts containing the Map and Slideshow blocks loaded and displayed fine.

Theme: Twenty Seventeen

pento commented:
@jeherve:

> It works well for the Map block, but does not seem to solve the issue for the Slideshow block on my end.

The Slideshow block appears fixed in my testing, I've tried a few variations, but haven't been able to make it break. Could you post steps to reproduce, theme you're using, and any JS console errors you see?

> Should we update our use of `@wordpress/dom-ready` for all the block that were touched in [d0453da](https://github.com/Automattic/jetpack/commit/d0453da32b31f2657020892fe7464d9f8f481087)?

You're referring to the `typeof window !== 'undefined'` checks? I wasn't quite sure why they're done, maybe in case the blocks are ever rendered server side? @simison, do you recall why those checks are there?

I've had a look through Gutenberg, and I couldn't see any such checks, or documentation that referred to them, either.

simison commented:
Blocks aren't server-side rendered but another environment needing those could be tests.

I often add the window checks in case since I've been burnt several times for not having them and someone being in trouble because of it. Could be they're unnecessary here.

jeherve commented:
> The Slideshow block appears fixed in my testing, I've tried a few variations, but haven't been able to make it break. Could you post steps to reproduce, theme you're using, and any JS console errors you see?

I'm using Twenty Nineteen. Right now both blocks appear not to work. You can see them on [this site](https://jeherve.com/news/), on page 4 when scrolling. The 2 posts are these ones:
https://jeherve.com/the-lady-in-the-lake/
https://jeherve.com/wanheda-1/

I don't see any console errors in Chrome, where I am testing. The blocks work well in the individual posts. Maybe worth noting is that the posts page is not the home page, as you can see. Also worth noting that there is a working map and a working slideshow on page 1 of the posts page.

pento committed:
Re-send already loaded block scripts, every time.

pento commented:
That was a good bug, thanks @jeherve! 😄

The bug was caused by a block script already having been loaded, the infinite scroll load won't send it again. Block scripts only run once on page load, however, so they won't run their init functionality for blocks added in subsequent infinite scroll loads.

39542b is a basic fix, but it currently re-renders all blocks on the page, which is a sub-optimal experience. I'll make it work a bit nicer.

pento committed:
Skip rendering blocks that have already been rendered.

pento committed:
Add a data-jetpack-block-initialized check to blocks that don't use FrontendManagement.

pento commented:
These are the blocks that do things in their `view.js` (other than load styles): `mailchimp`, `map`, `podcast-player`, `recurring-payments`, `repeat-visitor`, `revue`, `slideshow`, `tiled-gallery`.

Of those, `map` uses `FrontendManagement`, which I added a fix for in 
226d293. `repeat-visitor` doesn't touch blocks that have already been loaded.

The rest have custom loaders. I briefly considered re-writing them to use `FrontendManagement`, but given the wide variety of loaders, that seemed to be getting way out of scope for this PR. Instead, I've implemented simple checks in each of them, in the style of the check in `FrontendManagement`.

scruffian commented:
> The rest have custom loaders. I briefly considered re-writing them to use FrontendManagement, but given the wide variety of loaders, that seemed to be getting way out of scope for this PR.

Do you think it's worth adding an issue to the backlog so we can go back and tidy this up? It might prevent us having more bugs in the future....

scruffian commented:
I'm still having problems with this on TwentyNineteen. Map and Slideshow blocks don't load...

simison commented:
> Map and Slideshow blocks don't load...

I think both rely on external library loaded dynamically vs bundling it in (Mapbox & Swiper)

pento commented:
@scruffian: I haven't been able to reproduce this, it seems to be working fine in Twenty Nineteen for me. Are the blocks not loading on the front page, or only on an infinite scroll loaded page?

scruffian commented:
> Are the blocks not loading on the front page, or only on an infinite scroll loaded page?

Only on an infinite scroll page. I'm running it on localhost if that makes a difference?

pento commented:
I'm testing on `localhost`, too. To check the easy stuff first, do you have `SCRIPT_DEBUG` enabled, and the block JS files have been built?

Does your debugger show network requests for the respective block `view.js` files?

If you're not seeing requests at all, could you check ifthe JSON blob returned in the `/?infinity=scrolling` request include entries for the block `view.js` files?

scruffian committed:
Merge branch 'master' into fix/15025-map-block-infinite-scroll

scruffian commented:
I dont know what was going wrong before, but it's all working good in Twenty Nineteen now. Sorry about that. @jeherve hopefully this works for you now too...

kraftbj commented on extensions/blocks/map/view.js:
```diff
@@ -7,29 +12,28 @@ import { settings } from './settings.js';
+	const frontendManagement = new FrontendManagement();
+	// Add apiKey to attibutes so FrontendManagement knows about it.
+	// It is dynamically being added on the php side.
+	// So that it can be updated accross all the map blocks at the same time.
```
```suggestion
	// So that it can be updated across all the map blocks at the same time.
```

kraftbj commented on modules/infinite-scroll/infinity.php:
```diff
@@ -1056,7 +1056,19 @@ function filter_infinite_scroll_results( $results, $query_args, $wp_query ) {
+				$wp_scripts->done,
+				function ( $script_name ) use ( $initial_scripts ) {
+					// Jetpack block scripts should always be sent, even if they've been
+					// sent before. These scripts only run once on when loadeed, they don't
```
```suggestion
					// sent before. These scripts only run once on when loaded, they don't
```

kraftbj committed:
Update extensions/blocks/map/view.js

kraftbj reviewed (dismissed):
Works in my tests. I committed a couple fixes for typos.

IS is pretty out of sync with WP.com, so proceed slowly. I'd like to get them back in sync soon, so if you can apply this when you merge, great, but don't let it stop you if you don't.

jeherve committed:
Fix typo

Co-authored-by: Brandon Kraft <public@brandonkraft.com>

jeherve reviewed (approved):
This works well in my tests. Should be good to merge!

pento commented:
D43744-code merges all of infinite scroll back to dotcom. It seems to be working pretty well in my initial testing.
